### PR TITLE
A very rough security and privacy questionnaire

### PR DESCRIPTION
**Pull request by martinthomson**
I'm trying to be as frank and direct about risks and costs as possible.  It's very easy to spin this one way or other, with the goal of making the proposal seem more attractive.  Though very common, I find that to be disrespectful of readers.  This is already a hard enough topic to grok without the answers you need being obfuscated in that way.

### CODE REVIEW

File: sec-priv-q.md [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+agent in site-specific storage; subsequent requests result in the same value
+being returned.
+
+
+## Key Challenges
+
+There two major components to this work that require special attention from a
+privacy perspective:
+
+1. This proposal uses information - match keys - that might be used to perform
```
[the remark this answers is not in the document]
Comment (martinthomson) [reply sec-priv-q.md:65]: Only if the protections we apply fail.  I'll add that.

File: sec-priv-q.md [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+defined MPC protocol, the value of the match key is never revealed to any
+entity.
+
+IPA defines an MPC protocol that provides sites with an aggregated report that
+can be used for attribution.  The MPC system chosen guarantees *malicious
+security* in the *two out of three honest majority* setting.  In this setting,
+both privacy for inputs and security against tampering are guaranteed as long as
+two of the three helper parties faithfully executes the protocol.  It takes two
+corrupted or malicious helpers that actively work together to access the input
+data provided about users or to spoil the output of the computation.
```
[the remark this answers is not in the document]
Comment (martinthomson) [reply sec-priv-q.md]: True.  I'll say that as well.  Two defectors can spoil the output, one can only stop the output from being produced.

File: sec-priv-q.md [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+through all of the requirements and constraints in this area, such that we might
+infer the correct design.
+
+### Do the features in your specification expose information about the
+    underlying platform to origins?
+
+No.  However, if a compatible system is developed in the underlying platform for
+the management of match keys, user agent implementations might move to unify the
+two sources of match keys.  This would enable attribution that crosses from the
+web into applications.  No such mechanism currently exists.
```
Comment (benjaminsavage) [inline sec-priv-q.md:175]: well... it kind of does with PCM: https://webkit.org/blog/12042/pcm-for-in-app-direct-response-advertising/

File: sec-priv-q.md [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+unlinkable encryption is generated if site storage is cleared (and the
+underlying value might also need to be randomized too, though this question
+still hasn't been completely resolved; see above).
+
+### How does this specification distinguish between behavior in first-party and
+    third-party contexts?
+
+Information provided by the API is only provided to the top-level browsing
+context (a.k.a., the "first-party context").  Other sites that contribute to the
+content of a web page only receive a copy of the value provided to the top-level
```
Comment (benjaminsavage) [inline sec-priv-q.md:216]: I think you only need one only =).

File: sec-priv-q.md [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+corrupted or malicious helpers that actively work together to access the input
+data provided about users or to spoil the output of the computation.  Any attack
+by a single helper party can only cause the protocol to abort.
+
+User agent implementers, on behalf of their users, need to select helper parties
+that can be trusted to correctly execute the protocol and to only reveal the
+output of any processing they perform.  In addition to the cryptographic
+protections built into the MPC protocol design, the specification recommends
+that helper parties be bound by contractual terms with user agent
+implementations to execute the protocol correctly.
```
Comment (benjaminsavage) [inline sec-priv-q.md:53]: Is it worth noting that we have considered adding additional technical mitigations to make collusion more difficult, i.e. defence in depth through making the helpers run the protocol in a TEE?

File: sec-priv-q.md [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+   at which sites gain this information.
+
+Any conclusions about the privacy properties of the API will depend on an
+assessment of the adequacy of the design for both types of protection.
+
+
+## Standard Questions
+
+### What information might this feature expose to Web sites or other parties,
+    and for what purposes is that exposure necessary?
```
Comment (eriktaubeneck) [inline sec-priv-q.md:89]: This part of the question isn't addressed yet. Should be something to the effect of "learning aggregate information about activity that is specifically designed to be cross site, most commonly, paid advertising"

File: sec-priv-q.md (hunk 3 above)
Comment (martinthomson) [reply sec-priv-q.md:175]: I'm talking about a system that is compatible with IPA.  I don't think PCM really does that.  An analogous system exists, sure.

**Comment by martinthomson** [comment]
Added text covering the "necessity" question I missed, as @eriktaubeneck pointed out.

**Review by benjaminsavage — APPROVED** [review]
Looks good! We can follow up with additional PRs if we find anything was missed.

File: sec-priv-q.md [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+two of the three helper parties faithfully executes the protocol.  It takes two
+corrupted or malicious helpers that actively work together to access the input
+data provided about users or to spoil the output of the computation.
+
+User agent implementers, on behalf of their users, need to select helper parties
+that can be trusted to correctly execute the protocol and to only reveal the
+output of any processing they perform.  In addition to the cryptographic
+protections built into the MPC protocol design, the specification recommends
+that helper parties be bound by contractual terms with user agent
+implementations to execute the protocol correctly.
```
Comment (bmcase) [inline sec-priv-q.md:53]: "user agent implementations" -> implementers

File: sec-priv-q.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (bmcase) [reply sec-priv-q.md]: yeah, maybe "that could be used to perform cross-site tracking if the protections we apply fail."

File: sec-priv-q.md [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+    Private Browsing or Incognito mode?
+
+This question is not entirely resolved.  User agents that want maximum privacy
+protection - which seems likely in this particular case - can randomize the
+match key that is provided to sites.  This ensures that attribution is not
+possible based on the information that is provided.  This behaviour is
+indistinguishable to a site from use of a real match key and so does not offer
+sites a reason to discriminate against users.
+
+The same strategy can be employed for users who opt out of use of the API.
```
Comment (bmcase) [inline sec-priv-q.md:246]: Maybe worth adding a bit more on when to generate random matchkeys and when to cache their encryptions. 

An opt-out user agent will generate a random matchkey (preferably in some large set of opt-out/dummy matchkeys) for each site that requests it and cache the encryption of it for an epoch.  Different sites will have different matchkeys so there won’t be a chance of attribution.  Caching for each site over an epoch is necessary to ensure that the opt-out choice is not visible to sites.

**Pull request merged**